Add ASPA Support

tanneberger opened the pull request:
## Introduction
This PR introduces compatibility for RTR version 2 and handling of ASPA RTR PDUs as specified in [draft-ietf-sidrops-8210bis-11, section 5.12](https://datatracker.ietf.org/doc/html/draft-ietf-sidrops-8210bis-11#name-aspa-pdu) and adds AS_PATH verification according to [draft-ietf-sidrops-aspa-verification-16, section 6](https://datatracker.ietf.org/doc/html/draft-ietf-sidrops-aspa-verification-16#name-as_path-verification).

## Overview
This PR adds a new module called `aspa` which defines a new `aspa_table` organizing validated Autonomous System Provider Authorization data received from an RPKI-RTR cache server.  The `rtr` module was enhanced to parse ASPA PDUs and update the `aspa_table` accordingly. Analogous to `rtr_mgr_validate` (prefixes/ROAs) and `rtr_mgr_get_spki`, we're introducing a new function `rtr_mgr_verify_as_path` for AS_PATH verification. `rtrclient` has gained a new option, `-a`, for printing ASPA data in addition to `-p` (prefixes) and `-k`.

## Detailed Design

#### Storing ASPA Records
[ASPA record](https://github.com/tanneberger/rtrlib/blob/master/rtrlib/aspa/aspa.h#L38)s are structs derived from [RTR ASPA PDU](https://github.com/tanneberger/rtrlib/blob/master/rtrlib/rtr/rtr_pdus.h#L128)s, each containing a customer ASN and a set of providers attested by the customer. The `aspa_table` consists of a linked list where each node holds a reference to an `rtr_socket` and an `aspa_array`⏤an ordered dynamic array of ASPA records. Each `aspa_array` is comprised of a contiguous block of memory of ASPA records ordered by their customer ASN, enabling faster lookups by binary searching the array of records.

#### Updating an ASPA Table
The ASPA table implements aggregated updating using an array of *add record* and *remove record* operations that will be performed sequentially on the ASPA table. This drastically decreases the number of iterations necessary on the existing `aspa_array` to 1. This PR implements two update mechanisms:
- **In-Place**: The `aspa_array` is manipulated in-place in a single iteration, reducing new memory allocations while blocking callers wanting to verify an AS_PATH (therefore needing read access to the table).
- **Swap-In**:  A new `aspa_array` is created, incorporating both existing records and simultaneously adding new records in a single iteration. While this approach enables non-blocking updates, it's got a slightly higher memory demand.

Details: [aspa_private.h](https://github.com/tanneberger/rtrlib/blob/master/rtrlib/aspa/aspa_private.h#L16)

## Alternatives Considered
#### Other Data Structures
- splaytree-4
- kbtree-4
- dynarray (only sorted by cas)
- binsearch (provider arrays and cas are sorted)

This plot shows how the different data structures compare in different benchmarks against each other:

#### AS_PATH Verification Algorithm Implementation
We tested several implementations of the verification algorithm: [netd-tud/IETF-ASPA/hackathon](https://github.com/netd-tud/IETF-ASPA/tree/master/ietf-hackathon). The algorithm integrated into this PR is an [optimized version](https://github.com/netd-tud/IETF-ASPA/blob/master/ietf-hackathon/optimizedZeroBased.py) of the draft algorithm which avoids checking any AS hop twice.

## Tests & Quality
This PR has test coverage for 
- ASPA PDUs,
- modifications and search on an `aspa_array`,
- updates to an `aspa_table` and rtr_sync with ASPA
- AS_PATH verification
- Interoperability tests with standard RTR-Cache Server implementations (Routinator) were successfully performed.

Additionally, we checked for memory leaks and other faulty memory access using `valgrind --leak-check=full`.

#### Stats
Based on ASRank relationships we estimate ~75K ASPA records inferred from BGP paths. As an upper limit we consider an ASPA record to take up 24 bytes of memory.

#### Update Mechanism Performance Measurements
We conducted performance tests on both update approaches:
- Using the **In-Place** approach, adding 75K ASPA records in single update function call took ~100 ms on a reasonably sized laptop.
- The **Swap-In** mechanism has almost no effect on the time read access is blocked due to the old array just being swapped out in exchange for the new one. 

**Ping:** @waehlisch @mroethke @fho

waehlisch commented:
@mroethke @fho do you have time to give some feedback on this PR? ASPA is getting momentum and it would be important to provide ASPA support in the RTRlib.

fho commented:
> @mroethke @fho do you have time to give some feedback on this PR? ASPA is getting momentum and it would be important to provide ASPA support in the RTRlib.

Yes, I'll have a look before end of next week.

fho commented on rtrlib/rtr_mgr.c:
```diff
@@ -292,16 +293,19 @@ int rtr_mgr_config_cmp_tommy(const void *a, const void *b)
 	return rtr_mgr_config_cmp(ar->group, br->group);
 }
 
+// TODO: Additional arguments trailing?
```
Is the TODO comment still needed? If yes, I suggest to clarify it, I don't understand it.

carl-tud replied:
Due to the added aspa update fp, we’re introducing a source-breaking change here.. The alternative would’ve been to add the aspa fp argument as the last parameter (at the expense of consistency, still source-breaking)

tanneberger replied:
We thought about it a lot how we could avoid breaking user space, but there was always at least one bitter pill to swallow. Question is now which bitter pill do you want @fho.

fho replied:
Difficult, it's not even clear to me how bad it would be to break compatibility.
rtrlib  still has major version 0, which allows downwards incompatible changes according to the [semver](https://semver.org/) spec.
It would be good to release version 1 some when and guarantee API compatibility. 

The following alternatives come to my mind, you probably also discussed them already:
- keep `rtr_mgr_init` as it is, introduce a `rtr_mgr_init1` or `rtr_mgr_init_with_aspa` function that does the same but also has the new apsa_update_fp parameter
- keep `rtr_mgr_init` as it is, add a new function that only sets  the `aspa_update_fp` (`rtr_mgr_register_aspa_callback()` or so).
For version 1 rtr_mgr_init could also be refactored to  set all update_fps via separate functions.

fho replied:
@carl-tud @tanneberger any thought regarding https://github.com/rtrlib/rtrlib/pull/285#discussion_r1526483840?

tanneberger replied:
good reminder - thank you, I actually built this a couple of weeks ago and just forgot to push it.

I added the `aspa_array_reserve` function which can pre-allocate memory and started using it at all places, where we know the size already before.

tanneberger replied:
@fho I personally would just change the signature of `rtr_mgr_init` and then create a new release (v0.9.0). 

- This gives people the option to sick with the current release and if there is anything critical, we update both versions until everybody migrated.

fho replied:
> @fho I personally would just change the signature of rtr_mgr_init and then create a new release (v0.9.0).

Then lets go with that! :-)

>Other idea is to do a proper refactor: - What is your opinion on this? :

I like that it introduces an API that could make it easier in the future to add more functionality while keeping the API downwards compatible. I don't know if more functionality that requires a `rtr_mgr_init` signature change will be added in the future though.
To fully initialize the rtrlib 3 additional function calls would always be needed. I wonder if there are use cases where you don't want to have the full functionality. If there aren't or only very rarely, it makes the initialization more cumbersome

tanneberger replied:
But I guess we still can add functions like: `rtr_mgr_init_all(aspa_fp, spki_fp, roa_fp);`

fho reviewed (approved):
The implementation looks great to me, well structured, simple to follow and understand. The PR description was very helpful, it made it easy for me to get into it.
I could not find any issues, only trivial things that could be improved.
I was not able to verify every detail but everything I looked up was aligned with the draft.

carl-tud commented on rtrlib/rtr/packets.c:
```diff
@@ -893,6 +897,48 @@ static int rtr_store_router_key_pdu(struct rtr_socket *rtr_socket, const void *p
+		void *tmp = lrtr_realloc(*array, *capacity * sizeof(struct pdu_aspa *));
+
+		if (!tmp) {
+			const char txt[] = "Realloc failed";
+
+			RTR_DBG("%s", txt);
+			rtr_send_error_pdu_from_host(rtr_socket, NULL, 0, INTERNAL_ERROR, txt, sizeof(txt));
+			rtr_change_socket_state(rtr_socket, RTR_ERROR_FATAL);
```
Completely agree. The existing implementation however is sending these internal error messages in other places, too (e.g. in ˋrtr_store_[pfx|spki]_pduˋ). I guess these have to be fixed, too.

tanneberger replied:
Agreed: this is nonsensical. Question is now should we fix it in this PR or a new one?

fho replied:
in another PR, it's a different issue

tanneberger replied:
ok

carl-tud commented:
Thanks, @fho. I don’t really have time this week, but I’ll address the changes you requested next week

tanneberger commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -0,0 +1,230 @@
+
+// MARK: - Manipulation
+
+static enum aspa_status aspa_array_reallocate(struct aspa_array *array)
```
We can totally give the user a create function which takes the initial size of the array. But later one we want to use this array as a highy level data structure (comparable to `std::vector<>`) where we don't need to worry about resizing, that's why we built this abstraction.

fho replied:
My assumption is that whenever more entries to the array are appended, the number of elements that are added is known beforehand. 
Therefore the array could be allocated with the exact size that is required.

This guarantees that:
1. Only 1 malloc operation is needed, with a dynamically resized array sometimes multiple ones will be done.
2. Only the exact amount of memory that is needed has been allocated

This is why a fixed size array seems superior in this usecase.
Am I missing something here?

tanneberger replied:
I added the `aspa_array_reserve` function which can pre-allocate memory and started using it at all places, where we know the size already before.  

(see last commit)

fho replied:
great :tada:

Could you elaborate on in which situations the size of the array is not known before and why? 
It's not clear to me and it would take me much more type to figure it out on my own. :-)
Thanks!

tanneberger commented on rtrlib/rtr/packets.c:
```diff
@@ -481,6 +388,26 @@ static bool rtr_pdu_check_size(const struct pdu_header *pdu)
+		// ASN is 4 bytes each
+		expected_size += asn_count * sizeof(aspa_pdu->provider_asns[0]);
+		if (aspa_pdu->len != expected_size) {
+			RTR_DBG1("PDU is too small to contain valid ASPA PDU!");
```
@fho if you look at the line above (L403) we calculate the expected size based on  the partially parsed telegram (`expected_size += asn_count * sizeof(aspa_pdu[0]->provider_asns[0])`). 

I changed the message to `The PDU is malformed, because the specified size doesn't match the real PDU size.`

tanneberger commented on rtrlib/rtr/packets.c:
```diff
@@ -995,19 +1041,397 @@ static int rtr_update_spki_table(struct rtr_socket *rtr_socket, struct spki_tabl
+
+	if (res == ASPA_SUCCESS) {
+		return RTR_SUCCESS;
+	} else {
```
done

tanneberger commented on rtrlib/aspa/aspa_array/aspa_array.h:
```diff
@@ -0,0 +1,106 @@
+ */
+
+#ifndef RTR_ASPA_DYN_ARRAY_H
+#define RTR_ASPA_DYN_ARRAY_H
```
done

tanneberger commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -0,0 +1,230 @@
+
+void aspa_array_free(struct aspa_array *array, bool free_provider_arrays)
+{
+	// if the array is null just return
```
done

tanneberger commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -0,0 +1,230 @@
+	// allocation the chunk of memory of the provider as numbers
+	struct aspa_record *data_field = lrtr_malloc(sizeof(struct aspa_record) * default_initial_size);
+
+	// malloc failed so returning an error
```
done

tanneberger commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -0,0 +1,230 @@
+{
+	const size_t default_initial_size = 128;
+
+	// allocation the chunk of memory of the provider as numbers
```
done

tanneberger commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -0,0 +1,230 @@
+	if (!data_field)
+		return ASPA_ERROR;
+
+	// allocating the aspa_record itself
```
done

tanneberger commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -0,0 +1,230 @@
+	// allocating the aspa_record itself
+	struct aspa_array *array = lrtr_malloc(sizeof(struct aspa_array));
+
+	// malloc for aspa_record failed hence we return an error
```
done

tanneberger commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -0,0 +1,230 @@
+			}
+		}
+
+		// freeing the data
```
done

tanneberger commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -0,0 +1,230 @@
+		lrtr_free(array->data);
+	}
+
+	// freeing the array itself
```
done

tanneberger commented on rtrlib/aspa/aspa_verification.c:
```diff
@@ -0,0 +1,308 @@
+{
+	bool customer_found = false;
+
+	for (struct aspa_store_node *node = aspa_table->store; !!node; node = node->next) {
```
yeah changed it to `node != NULL` 

done

tanneberger commented on rtrlib/aspa/aspa_verification.c:
```diff
@@ -0,0 +1,308 @@
+	}
+
+	pthread_rwlock_unlock(&aspa_table->lock);
+	return ASPA_AS_PATH_UNKNOWN;
```
done

tanneberger commented:
@fho Did you take a look at the new changes yet?

fho commented:
> @fho Did you take a look at the new changes yet?

No, I was still waiting for an answer regarding the changed rtr_mgr_init signature + compatibility (https://github.com/rtrlib/rtrlib/pull/285#discussion_r1526483840)

fho commented:
@tanneberger I think there is a confusion because you replied on another thread, that I did not meant. (or you reply was not related to my last comment).
The rtr_mgr_init discussion is still open (https://github.com/rtrlib/rtrlib/pull/285#discussion_r1526483840).
That you changed the array allocation I saw (https://github.com/rtrlib/rtrlib/pull/285#discussion_r1587914072)

tanneberger commented:
@fho I quickly implemented by suggestion, see last commit.

waehlisch commented:
CI: rerun

waehlisch commented:
can you rebase?

tanneberger commented:
CI: rerun

fho commented:
Let's merge it as soon as @tanneberger gives the final go.

fho commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -228,3 +222,35 @@ struct aspa_record *aspa_array_search(struct aspa_array *array, uint32_t custome
+	// coping the data from the old array into the new one
+	if (memcpy(data, array->data, array->size * sizeof(struct aspa_record)) == NULL) {
+		return ASPA_ERROR;
+	}
```
Can't we use realloc/reallocarray here instead of malloc+ memcpy?

tanneberger replied:
done

fho commented on rtrlib/aspa/aspa_array/aspa_array.c:
```diff
@@ -0,0 +1,256 @@
+	return aspa_array_insert_internal(array, array->size, record, copy_providers);
+}
+
+enum aspa_status aspa_array_remove(struct aspa_array *array, size_t index, bool free_providers)
```
If I understand it correctly the allocated memory for the array is currently not reduced when only remove operations are done.
Why is this not done and needed?
Can't this lead to a situation where the array uses much memory then actually needed for a long period of time?

tanneberger replied:
This could be, but I guess is unlikely. 

Do you think it is necessary to scale it down again?

fho replied:
To me it is still unclear why a dynamic resizing array is a better choice then a fixed size array, that would only allocate as much space as needed.

> Do you think it is necessary to scale it down again?

As I understand, in the current implementation we always allocate `2 * old_capacity` elements when more space is needed.
Therefore we can end up with a lot of unused space. For example, when the array holds 2048 elements and we add/remove 1 additional one we end up with having space for 4096 elements allocated unnecessarily for a long time.
In the worst scenario this could even lead to that aspa_array holds more unused memory then rtrlib uses effectively. I think we should make it more efficient and prevent that aspa_array can end up holding so much unused memory.

tanneberger replied:
@fho I implemented also now a scale down now in the `aspa_array_remove` function, so the capacity is never larger then twice the size of the array.

carl-tud replied:
@fho Out of interest, wouldn't have lowering the factor also solved this? I was worried about the number of allocations, but again, this depends on the number and size of updates we get..

tanneberger replied:
Its a trade of. The lower the factor, the more often we need to reallocate the higher the factor, the more wasted memory we potentially have.

fho replied:
> @fho Out of interest, wouldn't have lowering the factor also solved this? I was worried about the number of allocations, but again, this depends on the number and size of updates we get..

Also with a lower factor we might end up in the situation that we add a lot of records to the array, then remove the majority again and never need the memory again:

Exaggeration:
- we add 10.000 aspa records,
- we remove 9.000 aspa records, never need more space then for 1000 elements again,
=> we have allocated unnecessarily space for 9.000 elements until the program terminates


The factor of 2 also seems to big to me, the bigger the array grows the more unlikely it becomes that we ever use the reserved space.

tanneberger replied:
its now a linear scaling with offset 1000.

tanneberger commented:
@fho I also had to change the user interface: see https://github.com/rtrlib/rtrlib/pull/285/commits/0dbb12331da38bd404b814f7d8b2bc1f18f2370f

-> `rtr_mgr_init` also creates the sockets and copies the pointers to the (prefix/spki/aspa) table into the socket struct. This is problematic because at this point the various tables have not yet been initialized and the socket therefore only has null pointers.
-> By creating a new function `rtr_mgr_setup_sockets` which is called by the user after all correct table pointers have been copied to the sockets

I also added more validation to handle the new scenario that the user only initialized some tables. I added checks in the validation functions and in the update functions: See this commit: https://github.com/rtrlib/rtrlib/pull/285/commits/7a2b9510cd16f436167f5d9b0c82e3f10b5ab95c

fho commented:
Link to `aspa_array_reserve` discussion: https://github.com/rtrlib/rtrlib/commit/6eed5ee4f3334d2ee3195e6a75ac69790ede7d52#r142285681

I had a hard time to find it in the UI, GitHub has hidden after a small bubble symbol next to the commit
